src/librustc/ty/subst.rs

```diff
@@ -202,19 +179,25 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
     where FR: FnMut(&ty::RegionParameterDef, &[Kind<'tcx>]) -> &'tcx ty::Region,
           FT: FnMut(&ty::TypeParameterDef<'tcx>, &[Kind<'tcx>]) -> Ty<'tcx> {
         let defs = tcx.lookup_generics(def_id);
-        let mut substs = Vec::with_capacity(defs.count());
-
-        Substs::fill_item(&mut substs, tcx, defs, &mut mk_region, &mut mk_type);
-
-        Substs::new(tcx, substs)
+        let count = defs.count();
+        if count <= 8 {
```

Comment:
I'd be fine with having `Extend` on `MaybeSmallVec8` (we can rename it and hide the other one?).
Just need to create it `with_capacity(count)` and have it *never* upgrade from `Small` to `Large`.